File: src/providers/pty.py
"""PTY-based usage providers wrapping existing collectors."""

import time
from typing import Optional

from .base import DataSource, FetchResult, UsageProvider, PersistentUsageProvider
from ..collectors.claude import ClaudeEphemeralCollector, ClaudePersistentCollector
from ..collectors.codex import CodexEphemeralCollector, CodexPersistentCollector


class ClaudePTYProvider(UsageProvider):
    """Ephemeral PTY-based Claude usage provider."""

    def __init__(self):
        super().__init__()
        self.source_type = DataSource.PTY
        self._collector = ClaudeEphemeralCollector()

    def is_available(self) -> bool:
        """PTY is always available (depends on CLI being in PATH)."""
        return True

    def fetch(self) -> FetchResult:
        """Fetch usage data via ephemeral tmux session."""
        timestamp = time.time()

        try:
            metrics = self._collector.collect()

            # Check if we got valid data
            if not metrics or all(v.get('used_pct', -1) == 0 for v in metrics.values() if isinstance(v, dict)):
                # Empty metrics likely means stale session
                return FetchResult(
                    metrics=metrics,
                    source=self.source_type,
                    timestamp=timestamp,
                    error="Empty metrics returned (possibly stale session)",
                    stale=True
                )

            return FetchResult(
                metrics=metrics,
                source=self.source_type,
                timestamp=timestamp,
                error=None
            )

        except Exception as e:
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error=f"PTY collection failed: {str(e)}"
            )


class ClaudePersistentPTYProvider(PersistentUsageProvider):
    """Persistent PTY-based Claude usage provider."""

    def __init__(self):
        super().__init__()
        self.source_type = DataSource.PTY
        self._collector = ClaudePersistentCollector()

    def is_available(self) -> bool:
        """PTY is always available (depends on CLI being in PATH)."""
        return True

    def start(self) -> FetchResult:
        """Start persistent tmux session and get initial data."""
        timestamp = time.time()

        try:
            metrics = self._collector.start()

            return FetchResult(
                metrics=metrics,
                source=self.source_type,
                timestamp=timestamp,
                error=None
            )

        except Exception as e:
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error=f"PTY start failed: {str(e)}"
            )

    def refresh(self) -> FetchResult:
        """Refresh data from existing tmux session."""
        timestamp = time.time()

        try:
            metrics = self._collector.refresh()

            # Check if we got valid data
            if not metrics or all(v.get('used_pct', -1) == 0 for v in metrics.values() if isinstance(v, dict)):
                return FetchResult(
                    metrics=metrics,
                    source=self.source_type,
                    timestamp=timestamp,
                    error="Empty metrics returned (possibly stale session)",
                    stale=True
                )

            return FetchResult(
                metrics=metrics,
                source=self.source_type,
                timestamp=timestamp,
                error=None
            )

        except Exception as e:
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error=f"PTY refresh failed: {str(e)}"
            )

    def stop(self):
        """Stop persistent tmux session."""
        try:
            self._collector.stop()
        except Exception:
            # Best effort cleanup
            pass

    def fetch(self) -> FetchResult:
        """Fetch is not supported for persistent provider, use refresh instead."""
        return FetchResult(
            metrics=None,
            source=self.source_type,
            timestamp=time.time(),
            error="Use start() and refresh() for persistent provider"
        )


class CodexPTYProvider(UsageProvider):
    """Ephemeral PTY-based Codex usage provider."""

    def __init__(self):
        super().__init__()
        self.source_type = DataSource.PTY
        self._collector = CodexEphemeralCollector()

    def is_available(self) -> bool:
        """PTY is always available (depends on CLI being in PATH)."""
        return True

    def fetch(self) -> FetchResult:
        """Fetch usage data via ephemeral tmux session."""
        timestamp = time.time()

        try:
            metrics = self._collector.collect()

            # Check if we got valid data
            if not metrics or all(v.get('used_pct', -1) == 0 for v in metrics.values() if isinstance(v, dict)):
                return FetchResult(
                    metrics=metrics,
                    source=self.source_type,
                    timestamp=timestamp,
                    error="Empty metrics returned (possibly stale session)",
                    stale=True
                )

            return FetchResult(
                metrics=metrics,
                source=self.source_type,
                timestamp=timestamp,
                error=None
            )

        except Exception as e:
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error=f"PTY collection failed: {str(e)}"
            )


class CodexPersistentPTYProvider(PersistentUsageProvider):
    """Persistent PTY-based Codex usage provider."""

    def __init__(self):
        super().__init__()
        self.source_type = DataSource.PTY
        self._collector = CodexPersistentCollector()

    def is_available(self) -> bool:
        """PTY is always available (depends on CLI being in PATH)."""
        return True

    def start(self) -> FetchResult:
        """Start persistent tmux session and get initial data."""
        timestamp = time.time()

        try:
            metrics = self._collector.start()

            return FetchResult(
                metrics=metrics,
                source=self.source_type,
                timestamp=timestamp,
                error=None
            )

        except Exception as e:
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error=f"PTY start failed: {str(e)}"
            )

    def refresh(self) -> FetchResult:
        """Refresh data from existing tmux session."""
        timestamp = time.time()

        try:
            metrics = self._collector.refresh()

            # Check if we got valid data
            if not metrics or all(v.get('used_pct', -1) == 0 for v in metrics.values() if isinstance(v, dict)):
                return FetchResult(
                    metrics=metrics,
                    source=self.source_type,
                    timestamp=timestamp,
                    error="Empty metrics returned (possibly stale session)",
                    stale=True
                )

            return FetchResult(
                metrics=metrics,
                source=self.source_type,
                timestamp=timestamp,
                error=None
            )

        except Exception as e:
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error=f"PTY refresh failed: {str(e)}"
            )

    def stop(self):
        """Stop persistent tmux session."""
        try:
            self._collector.stop()
        except Exception:
            # Best effort cleanup
            pass

    def fetch(self) -> FetchResult:
        """Fetch is not supported for persistent provider, use refresh instead."""
        return FetchResult(
            metrics=None,
            source=self.source_type,
            timestamp=time.time(),
            error="Use start() and refresh() for persistent provider"
        )

File: packages/core/src/providers/pty.ts
/**
 * PTY-based usage providers wrapping collectors.
 * Port of src/providers/pty.py
 */

import { DataSource } from "../types.js";
import type { FetchResult, UsageProvider, PersistentUsageProvider, MetricsDict } from "../types.js";
import { ClaudeEphemeralCollector, ClaudePersistentCollector } from "../collectors/claude.js";
import { CodexEphemeralCollector, CodexPersistentCollector } from "../collectors/codex.js";

/** Check if metrics are genuinely missing or came entirely from fallbacks */
function isLikelyStale(metrics: MetricsDict | null): boolean {
  if (!metrics) return true;
  // Check if there are any actual metric entries (not just subscription_type)
  const metricEntries = Object.entries(metrics).filter(([key]) => key !== "subscription_type" && key !== "__parsed");
  if (metricEntries.length === 0) return true;
  // If parser flagged that no regex matched, all values are from applyFallbacks()
  if ("__parsed" in metrics && (metrics as any).__parsed === false) return true;
  return false;
}

/** Ephemeral PTY-based Claude usage provider */
export class ClaudePTYProvider implements UsageProvider {
  name = "ClaudePTYProvider";
  sourceType = DataSource.PTY;

  private _collector = new ClaudeEphemeralCollector();

  isAvailable(): boolean {
    return true;
  }

  async fetch(): Promise<FetchResult> {
    const timestamp = Date.now() / 1000;

    try {
      const metrics = await this._collector.collect();

      if (isLikelyStale(metrics)) {
        return {
          metrics,
          source: this.sourceType,
          timestamp,
          error: "Empty metrics returned (possibly stale session)",
          stale: true,
        };
      }

      return { metrics, source: this.sourceType, timestamp, error: null, stale: false };
    } catch (e) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: `PTY collection failed: ${e instanceof Error ? e.message : String(e)}`,
        stale: false,
      };
    }
  }
}

/** Persistent PTY-based Claude usage provider */
export class ClaudePersistentPTYProvider implements PersistentUsageProvider {
  name = "ClaudePersistentPTYProvider";
  sourceType = DataSource.PTY;

  private _collector = new ClaudePersistentCollector();

  isAvailable(): boolean {
    return true;
  }

  async start(): Promise<FetchResult> {
    const timestamp = Date.now() / 1000;
    try {
      const metrics = await this._collector.start();

      if (isLikelyStale(metrics)) {
        return {
          metrics,
          source: this.sourceType,
          timestamp,
          error: "Empty metrics returned (possibly stale session)",
          stale: true,
        };
      }

      return { metrics, source: this.sourceType, timestamp, error: null, stale: false };
    } catch (e) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: `PTY start failed: ${e instanceof Error ? e.message : String(e)}`,
        stale: false,
      };
    }
  }

  async refresh(): Promise<FetchResult> {
    const timestamp = Date.now() / 1000;
    try {
      const metrics = await this._collector.refresh();

      if (isLikelyStale(metrics)) {
        return {
          metrics,
          source: this.sourceType,
          timestamp,
          error: "Empty metrics returned (possibly stale session)",
          stale: true,
        };
      }

      return { metrics, source: this.sourceType, timestamp, error: null, stale: false };
    } catch (e) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: `PTY refresh failed: ${e instanceof Error ? e.message : String(e)}`,
        stale: false,
      };
    }
  }

  async stop(): Promise<void> {
    try {
      await this._collector.stop();
    } catch {
      // Best effort cleanup
    }
  }

  async fetch(): Promise<FetchResult> {
    return {
      metrics: null,
      source: this.sourceType,
      timestamp: Date.now() / 1000,
      error: "Use start() and refresh() for persistent provider",
      stale: false,
    };
  }
}

/** Ephemeral PTY-based Codex usage provider */
export class CodexPTYProvider implements UsageProvider {
  name = "CodexPTYProvider";
  sourceType = DataSource.PTY;

  private _collector = new CodexEphemeralCollector();

  isAvailable(): boolean {
    return true;
  }

  async fetch(): Promise<FetchResult> {
    const timestamp = Date.now() / 1000;
    try {
      const metrics = await this._collector.collect();

      if (isLikelyStale(metrics)) {
        return {
          metrics,
          source: this.sourceType,
          timestamp,
          error: "Empty metrics returned (possibly stale session)",
          stale: true,
        };
      }

      return { metrics, source: this.sourceType, timestamp, error: null, stale: false };
    } catch (e) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: `PTY collection failed: ${e instanceof Error ? e.message : String(e)}`,
        stale: false,
      };
    }
  }
}

/** Persistent PTY-based Codex usage provider */
export class CodexPersistentPTYProvider implements PersistentUsageProvider {
  name = "CodexPersistentPTYProvider";
  sourceType = DataSource.PTY;

  private _collector = new CodexPersistentCollector();

  isAvailable(): boolean {
    return true;
  }

  async start(): Promise<FetchResult> {
    const timestamp = Date.now() / 1000;
    try {
      const metrics = await this._collector.start();

      if (isLikelyStale(metrics)) {
        return {
          metrics,
          source: this.sourceType,
          timestamp,
          error: "Empty metrics returned (possibly stale session)",
          stale: true,
        };
      }

      return { metrics, source: this.sourceType, timestamp, error: null, stale: false };
    } catch (e) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: `PTY start failed: ${e instanceof Error ? e.message : String(e)}`,
        stale: false,
      };
    }
  }

  async refresh(): Promise<FetchResult> {
    const timestamp = Date.now() / 1000;
    try {
      const metrics = await this._collector.refresh();

      if (isLikelyStale(metrics)) {
        return {
          metrics,
          source: this.sourceType,
          timestamp,
          error: "Empty metrics returned (possibly stale session)",
          stale: true,
        };
      }

      return { metrics, source: this.sourceType, timestamp, error: null, stale: false };
    } catch (e) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: `PTY refresh failed: ${e instanceof Error ? e.message : String(e)}`,
        stale: false,
      };
    }
  }

  async stop(): Promise<void> {
    try {
      await this._collector.stop();
    } catch {
      // Best effort cleanup
    }
  }

  async fetch(): Promise<FetchResult> {
    return {
      metrics: null,
      source: this.sourceType,
      timestamp: Date.now() / 1000,
      error: "Use start() and refresh() for persistent provider",
      stale: false,
    };
  }
}
